[ios][docs] Update background modes configuration for iOS in multiple SDK pages

chrfalch opened the pull request:
# Why

The current documentation for `background-fetch`, `location`, `notification` and `task-manager` needed a few clarifications on how the background-part of these libraries are set up and configured. This is specifically related to the required `UIBackgroundModes` on iOS which has requirements for the background-execution in these packages.

There were also a wrong reference to use the key `processing` that had to be removed since it causes a runtime error due to a missing additional required identifier that is not necessary in the current structure.

Closes #32396
Closes #32220

# How

- Removed example in task-manager doc and moved this into each respective library (with reference from task-manager)
- Removed reference to the "processing" background mode since it is not required (it is required if using BGTaskScheduler which we aren't yet)
- Aligned documentation for each library to use same structure:
  - Added notes about Expo Go App support 
  - Added notes on how to configure using CNG and for non-CNG and native apps.
  - Removed wrong reference in README in the `background-fetch` package

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [X] Conforms with the [Documentation Writing Style Guide]

chrfalch committed:
[ios][docs] Update background modes configuration for iOS in multiple SDK pages

Closes #32396
Closes #32220

The old documentation for the background-fetch, location, notification and task-manager needed a few clarifications on how the background-part of these libraries are set up and configured.

- Removed example in task-manager doc and moved this into each respective library (with reference from task-manager)
- Removed reference to the "processing" background mode since it is not required (it is required if using BGTaskScheduler which we aren't yet)
- Aligned documentation for each library to use same structure
- Added notes about Expo Go App support
- Added notes on how to configure using CNG and for non-CNG and native apps.

chrfalch committed:
[docs] Updated changelog

behenate commented on packages/expo-notifications/CHANGELOG.md:
```diff
@@ -509,9 +511,11 @@ _This version does not introduce any user-facing changes._
 - Changed class responsible for handling Firebase events from `FirebaseMessagingService` to `.service.NotificationsService` on Android. ([#10558](https://github.com/expo/expo/pull/10558) by [@sjchmiela](https://github.com/sjchmiela))
 
   > Note that this change most probably will not affect you — it only affects projects that override `FirebaseMessagingService` to implement some custom handling logic.
+
```
```suggestion

```

behenate commented on packages/expo-notifications/CHANGELOG.md:
```diff
@@ -509,9 +511,11 @@ _This version does not introduce any user-facing changes._
 - Changed how you can override ways in which a notification is reinterpreted from a [`StatusBarNotification`](https://developer.android.com/reference/android/service/notification/StatusBarNotification) and in which a [`Notification`](https://developer.android.com/reference/android/app/Notification.html?hl=en) is built from defining an `expo.modules.notifications#NotificationsScoper` meta-data val ...
 
   > Note that this change most probably will not affect you — it only affects projects that override those methods to implement some custom handling logic.
+
```
```suggestion

```

behenate commented on packages/expo-notifications/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@
 
 ### 💡 Others
 
+- [docs] Update background modes configuration for iOS in multiple SDK pages ([#32647](https://github.com/expo/expo/pull/32647) by [@chrfalch](https://github.com/chrfalch))
+
```
```suggestion

```

chrfalch replied:
Not sure about this one, @behenate?

behenate commented on packages/expo-task-manager/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@
 
 ### 💡 Others
 
+- [docs] Update background modes configuration for iOS in multiple SDK pages ([#32647](https://github.com/expo/expo/pull/32647) by [@chrfalch](https://github.com/chrfalch))
```
We usually don't add changelog entries for docs changes, only for changes in the code of the package. This is because the docs are published separately from the package.
```suggestion
```

chrfalch replied:
Absolutely - was just following the suggestion from the bot!!! 😅

behenate commented on packages/expo-location/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@
 
 ### 💡 Others
 
+- [docs] Update background modes configuration for iOS in multiple SDK pages ([#32647](https://github.com/expo/expo/pull/32647) by [@chrfalch](https://github.com/chrfalch))
+
```
```suggestion
```

behenate commented on packages/expo-background-fetch/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@
 
 ### 💡 Others
 
+- [docs] Update background modes configuration for iOS in multiple SDK pages ([#32647](https://github.com/expo/expo/pull/32647) by [@chrfalch](https://github.com/chrfalch))
+
```
```suggestion
```

behenate commented on docs/pages/versions/v52.0.0/sdk/location.mdx:
````diff
@@ -99,7 +100,27 @@ Learn how to configure the native projects in the [installation instructions in
 </ConfigReactNative>
 
-### Background Location methods
+### Background location configuration&ensp;<PlatformTag platform="ios" />
+
+To be able to run background location on iOS, you need to add the `location` value to the `UIBackgroundModes` array in your app's **Info.plist** file.
+
+**If you're using [CNG](/workflow/continuous-native-generation/)** the required `UIBackgroundModes` configuration will be applied automatically the next time you run `npx expo prebuild`.
+
+<Collapsible summary="Configure UIBackgroundModes manually on iOS">
+
+If you're not using Continuous Native Generation ([CNG](/workflow/continuous-native-generation/)) or you're using a native iOS project, then you'll need to add the following to your **Expo.plist** file:
+
+```xml ios/project-name/Supporting/Expo.plist
+<key>UIBackgroundModes</key>
+  <array>
+    <string>location</string>
+  </array>
+</key>
+```
+
+</Collapsible>
+
+### Background location methods
````
We need all of the changes from the /docs/pages/versions/v52.0.0/sdk to also be in /docs/pages/versions/v52.0.0/unversioned

chrfalch replied:
Ok! Will update :) Thanks for pointing out.

chrfalch replied:
Guess you meant in `/docs/pages/versions/unversioned`?

chrfalch replied:
Done :)

chrfalch committed:
Update packages/expo-notifications/CHANGELOG.md

Co-authored-by: Wojciech Dróżdż <31368152+behenate@users.noreply.github.com>

chrfalch committed:
Update packages/expo-notifications/CHANGELOG.md

Co-authored-by: Wojciech Dróżdż <31368152+behenate@users.noreply.github.com>

chrfalch committed:
Update packages/expo-notifications/CHANGELOG.md

Co-authored-by: Wojciech Dróżdż <31368152+behenate@users.noreply.github.com>

chrfalch committed:
Update packages/expo-task-manager/CHANGELOG.md

Co-authored-by: Wojciech Dróżdż <31368152+behenate@users.noreply.github.com>

chrfalch committed:
Update packages/expo-location/CHANGELOG.md

Co-authored-by: Wojciech Dróżdż <31368152+behenate@users.noreply.github.com>

chrfalch committed:
Update packages/expo-background-fetch/CHANGELOG.md

Co-authored-by: Wojciech Dróżdż <31368152+behenate@users.noreply.github.com>

chrfalch committed:
[docs] added changes to unversioned docs as well as in sdk-52

Simek commented on packages/expo-task-manager/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
 
 ### 💡 Others
 
+
```
```suggestion
```

Simek commented on docs/pages/versions/v52.0.0/sdk/notifications.mdx:
```diff
@@ -471,7 +491,11 @@ For Android:
 For iOS:
 
-- In the array under the `ios.infoPlist.UIBackgroundModes` key in your [app config](/workflow/configuration/), add the values `remote-notification` and `processing`.
+- background mode must be specified in **Info.plist** file. See [Background notifications
```
```suggestion
- Background mode must be specified in **Info.plist** file. See [Background notifications
```

Simek commented on docs/pages/versions/v52.0.0/sdk/notifications.mdx:
```diff
@@ -471,7 +491,11 @@ For Android:
-- In the array under the `ios.infoPlist.UIBackgroundModes` key in your [app config](/workflow/configuration/), add the values `remote-notification` and `processing`.
+- background mode must be specified in **Info.plist** file. See [Background notifications
+  configuration](#background-notification-configuration).
+- You must use a [development
+  build](/develop/development-builds/introduction/) to use background notifications since it is not
+  supported in the Expo Go app.
```
```suggestion
- You must use a [development build](/develop/development-builds/introduction/) to use background notifications since it is not supported in the Expo Go app.
```

chrfalch committed:
Update packages/expo-task-manager/CHANGELOG.md

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

Simek commented on docs/pages/versions/unversioned/sdk/notifications.mdx:
```diff
@@ -471,7 +491,11 @@ For Android:
 For iOS:
 
-- In the array under the `ios.infoPlist.UIBackgroundModes` key in your [app config](/workflow/configuration/), add the values `remote-notification` and `processing`.
+- background mode must be specified in **Info.plist** file. See [Background notifications
+  configuration](#background-notification-configuration).
+- You must use a [development
+  build](/develop/development-builds/introduction/) to use background notifications since it is not
+  supported in the Expo Go app.
```
```suggestion
- Background mode must be specified in **Info.plist** file. See [Background notifications configuration](#background-notification-configuration).
- You must use a [development build](/develop/development-builds/introduction/) to use background notifications since it is not supported in the Expo Go app.
```

chrfalch replied:
Thanks, @simek! I probably have to manually make these updates to the sdk-52 version as well?

chrfalch committed:
Update docs/pages/versions/v52.0.0/sdk/notifications.mdx

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

chrfalch committed:
Update docs/pages/versions/v52.0.0/sdk/notifications.mdx

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/notifications.mdx

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

Simek commented:
We are repeating now the list of links to supported libraries by `TaskManager` twice within the same viewport, maybe we could rewrite the content to avoid that?

![Screenshot 2024-11-06 at 15 35 37](https://github.com/user-attachments/assets/501fe6ab-2f04-4b97-8a62-f682f05a729c)

Simek reviewed (approved):
Left few suggestions, mostly formatting ones, but overall this looks great, thanks for updating those guides! 👍

chrfalch commented:
> Left few suggestions, mostly formatting ones, but overall this looks great, thanks for updating those guides! 👍

Awesome feedback, @Simek - thanks, really appreciate it! Committed all the suggestions - will go another round on the list of libraries thing - agree that it looks a bit weird.

Simek commented:
Just an idea: maybe we can move the list of supported packages into their own subsection with a header, which then can be linked in the "Configure" section?

chrfalch committed:
[codereview] made sure sdk-52 and unversioned are the same

Updated line break

chrfalch commented:
> Just an idea: maybe we can move the list of supported packages into their own subsection with a header, which then can be linked in the "Configure" section?

Something like this?

<img width="998" alt="image" src="https://github.com/user-attachments/assets/96cbb292-9e82-4946-bc94-2ac104e5fd77">

We'll not be able to include links to both the main page and the configuration section for each package, but that might be ok.

chrfalch committed:
Updated links to packages using task manager

Simek reviewed (approved):
Thanks for the update, it LGTM! 👍 

Let's make sure the @amandeepmittal does not have any suggestion before shipping.

amandeepmittal commented on docs/pages/versions/unversioned/sdk/task-manager.mdx:
```diff
@@ -8,11 +8,13 @@ platforms: ['android', 'ios']
 import { SnackInline } from '~/ui/components/Snippet';
+import { PlatformTag } from '~/ui/components/Tag';
 
 `expo-task-manager` provides an API that allows you to manage long-running tasks, in particular those tasks that can run while your app is in the background. Some features of this library are used by other libraries under the hood. Here is a list of Expo SDK libraries that use `TaskManager`:
```
```suggestion
`expo-task-manager` provides an API that allows you to manage long-running tasks, in particular those tasks that can run while your app is in the background. Some features of this library are used by other libraries under the hood. Here is a list of Expo SDK libraries that use `TaskManager`.
```

Since the libraries have their own section, let's update the intro to remove the colon.

amandeepmittal commented on docs/pages/versions/unversioned/sdk/task-manager.mdx:
```diff
@@ -21,33 +23,13 @@ import { SnackInline } from '~/ui/components/Snippet';
+Standalone apps need some extra configuration: on iOS, each background feature requires a special key in `UIBackgroundModes` array in your **Info.plist** file.
 
-</ConfigReactNative>
+Read more about how to configure this in the documentation pages for each of the [libraries](#libraries-using-expo-taskmanager) that use `TaskManager`.
```
```suggestion
Read more about how to configure this in the reference for each of the [libraries](#libraries-using-expo-taskmanager) that use `TaskManager`.
```

amandeepmittal commented on docs/pages/versions/v52.0.0/sdk/task-manager.mdx:
```diff
@@ -8,11 +8,13 @@ platforms: ['android', 'ios']
 import { SnackInline } from '~/ui/components/Snippet';
+import { PlatformTag } from '~/ui/components/Tag';
 
 `expo-task-manager` provides an API that allows you to manage long-running tasks, in particular those tasks that can run while your app is in the background. Some features of this library are used by other libraries under the hood. Here is a list of Expo SDK libraries that use `TaskManager`:
```
```suggestion
`expo-task-manager` provides an API that allows you to manage long-running tasks, in particular those tasks that can run while your app is in the background. Some features of this library are used by other libraries under the hood. Here is a list of Expo SDK libraries that use `TaskManager`
```

amandeepmittal commented on docs/pages/versions/v52.0.0/sdk/task-manager.mdx:
```diff
@@ -21,33 +23,13 @@ import { SnackInline } from '~/ui/components/Snippet';
+Standalone apps need some extra configuration: on iOS, each background feature requires a special key in `UIBackgroundModes` array in your **Info.plist** file.
 
-</ConfigReactNative>
+Read more about how to configure this in the documentation pages for each of the [libraries](#libraries-using-expo-taskmanager) that use `TaskManager`.
```
```suggestion
Read more about how to configure this in the reference for each of the [libraries](#libraries-using-expo-taskmanager) that use `TaskManager`.
```

amandeepmittal reviewed (approved):
Left a few minor suggestions below.

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/task-manager.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/task-manager.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update docs/pages/versions/v52.0.0/sdk/task-manager.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update docs/pages/versions/v52.0.0/sdk/task-manager.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch commented:
Thanks, @amandeepmittal and @Simek :) I've asked @brentvatne to have a look as well to make sure I've understood the concepts of CNG and everything related.

brentvatne commented on docs/pages/versions/unversioned/sdk/background-fetch.mdx:
```diff
@@ -21,17 +21,31 @@ import { PlatformTags } from '~/ui/components/Tag';
+To be able to run background fetch tasks on iOS, you need to add the `fetch` value to the `UIBackgroundModes` array in your app's **Info.plist** file. This is required for background fetch to work properly.
 
-<ConfigReactNative>
+**If you're using [CNG](/workflow/continuous-native-generation/)** the required `UIBackgroundModes` configuration will be applied automatically the next time you run `npx expo prebuild`.
```
```suggestion
**If you're using [CNG](/workflow/continuous-native-generation/)**, the required `UIBackgroundModes` configuration will be applied automatically by prebuild.
```

slight preference for this wording because i want to avoid people misinterpreting that they need to run prebuild on their machine if they aren't doing that already (often folks will just use eas build)

brentvatne reviewed (approved):
this pattern looks good to me! thanks @chrfalch!

chrfalch committed:
[codereview] adjusted wording slightly

From brent:
"slight preference for this wording because i want to avoid people misinterpreting that they need to run prebuild on their machine if they aren't doing that already (often folks will just use eas build)"

Good point!

vonovak commented on docs/pages/versions/unversioned/sdk/notifications.mdx:
```diff
@@ -454,6 +454,26 @@ In the table above, whenever `NotificationResponseReceivedListener` is triggered
+
+To be able to run background notifications on iOS, you need to add the `remote-notification` value to the `UIBackgroundModes` array in your app's **Info.plist** file.
+
+**If you're using [CNG](/workflow/continuous-native-generation/)**, the required `UIBackgroundModes` configuration will be applied automatically by prebuild.
```
@chrfalch 

the next paragraph "Background notifications" previously contained this:

```
- In the array under the `ios.infoPlist.UIBackgroundModes` key in your [app config](/workflow/configuration/), add the values `remote-notification` and `processing`.
```

I believe that instruction is still valuable (?) - or maybe I don't understand what `If you're using, the required `UIBackgroundModes` configuration will be applied automatically by prebuild.` really means. How does this "automatic" configuration work?

chrfalch replied:
Hi! Yes, the line was fixed since It contained an error - `processing` has nothing to do with notifications - it is for the new `BGTaskScheduler` API (which could be used in a new `expo-background-task` package probably). 

"Automatic" means that when you're using CNG (continuous native generation) `eas:build` or `prebuild` will update your info.plist file for you - there are plugins in each of the library responsible for doing so. I had a few discussions with Brent about how we could "standardize" on a specific wording in our docs - which resulted in this.

vonovak replied:
then I believe (?) that the following should stay:

`- In the array under the `ios.infoPlist.UIBackgroundModes` key in your [app config](/workflow/configuration/), add the value `remote-notification``

What confuses me is that after reading the docs I feel I don't need to do _anything at all_ to set up background notifications on iOS if I'm using prebuild.

However, we can't / shouldn't automatically decide for users whether to add `remote-notification` to `UIBackgroundModes` - I believe the users should do that themselves.

The [config plugin for notifications](https://github.com/expo/expo/blob/main/packages/expo-notifications/plugin/src/withNotifications.ts#L8) does not add `remote-notification` to `UIBackgroundModes` this and neither does [the one in task manager](https://github.com/expo/expo/blob/db38d257a2aaacc8562f553e72f07d00bcb5538f/packages/expo-task-manager/plugin/src/withTaskManager.ts#L5) afaict.
